app/uk/gov/hmrc/transitmovementsauditing/services/AuditService.scala

```diff
@@ -50,10 +58,41 @@ trait AuditService {
   def send(auditType: AuditType, jsonStream: Payload)(implicit
     hc: HeaderCarrier
   ): EitherT[Future, AuditError, Unit]
+
+  def getAdditionalField(name: String, path: Seq[String], src: Source[ByteString, _]): Future[ParseResult[(String, String)]]
+
+  def getAdditionalFields(messageType: Option[MessageType], src: Source[ByteString, _]): EitherT[Future, ParseError, Seq[ParseResult[(String, String)]]]
 }
 
 @Singleton
-class AuditServiceImpl @Inject() (connector: AuditConnector)(implicit ec: ExecutionContext, materializer: Materializer) extends AuditService {
+class AuditServiceImpl @Inject() (connector: AuditConnector)(implicit ec: ExecutionContext, materializer: Materializer)
+    extends AuditService
+    with ElementPaths
+    with Logging {
+
+  def getAdditionalField(name: String, path: Seq[String], src: Source[ByteString, _]): Future[ParseResult[(String, String)]] =
+    src
+      .via(XmlParsing.parser)
+      .via(XmlParsers.extractElement(name, path))
+      .runWith(concatKeyValue)
+
+  def getAdditionalFields(messageType: Option[MessageType], src: Source[ByteString, _]): EitherT[Future, ParseError, Seq[ParseResult[(String, String)]]] =
+    EitherT {
+      messageType match {
+        case Some(value) =>
+          Future
+            .sequence(
+              elementPaths(value.messageCode).map {
+                row =>
+                  getAdditionalField(row._1, row._2, src) // node name and its path
+              }.toSeq
+            )
+            .map(
+              keyValuePairs => Right(keyValuePairs)
+            )
+        case None => Future.successful(Left(ParseError.NoElementFound(s"Unable to find $messageType")))
+      }
+    }
```

Comment:
These really should be in a `ParsingService` rather than the `AuditService` -- the latter is about sending the Json payload, the former is about pulling stuff out of XML. We should keep them separate.